enc_bootloader/enc_bootloader.c

```diff
@@ -0,0 +1,202 @@
+        uint8_t div = 1;
+        uint32_t drives = 0x7777;
+        while (div < 4) {
+            rosc_hw->div = div | ROSC_DIV_VALUE_PASS;
+            rosc_hw->freqa = (ROSC_FREQA_PASSWD_VALUE_PASS << ROSC_FREQA_PASSWD_LSB) |
+                    drives |
+                    ROSC_FREQA_DS1_RANDOM_BITS | ROSC_FREQA_DS0_RANDOM_BITS; // enable randomisation
+            
+            // Wait for ROSC to be stable
+            while(!(rosc_hw->status & ROSC_STATUS_STABLE_BITS)) {
+                tight_loop_contents();
+            }
+
+            if (frequency_count_khz(CLOCKS_FC0_SRC_VALUE_ROSC_CLKSRC_PH) < clk_khz) {
+                break;
+            }
+
+            if (!drives) div++;
+            drives = drives ? 0x0000 : 0x7777;
+        }
+    }
+#endif // XOSC_CALIBRATION
+    // CLK SYS = ROSC directly, as it's running slowly enough
+    clock_configure_int_divider(clk_sys,
+                    CLOCKS_CLK_SYS_CTRL_SRC_VALUE_CLKSRC_CLK_SYS_AUX,
+                    CLOCKS_CLK_SYS_CTRL_AUXSRC_VALUE_ROSC_CLKSRC,
+                    ROSC_HZ,    // this doesn't have to be accurate
+                    1);
+
+    // Configure other clocks - none of these need to be accurate
+
+    // CLK_REF = ROSC / OTHER_CLK_DIV - this and other clocks aren't really used, so just need to be set to a low enough frequency
+    clock_configure_int_divider(clk_ref,
+                    CLOCKS_CLK_REF_CTRL_SRC_VALUE_ROSC_CLKSRC_PH,
+                    0,
+                    ROSC_HZ,
+                    OTHER_CLK_DIV);
+
+    // CLK USB (not used)
+    clock_configure_int_divider(clk_usb,
+                    0, // No GLMUX
+                    CLOCKS_CLK_USB_CTRL_AUXSRC_VALUE_ROSC_CLKSRC_PH,
+                    ROSC_HZ,
+                    OTHER_CLK_DIV);
+
+    // CLK ADC (not used)
+    clock_configure_int_divider(clk_adc,
+                    0, // No GLMUX
+                    CLOCKS_CLK_ADC_CTRL_AUXSRC_VALUE_ROSC_CLKSRC_PH,
+                    ROSC_HZ,
+                    OTHER_CLK_DIV);
+
+    // CLK PERI Used as reference clock for UART and SPI serial. (not used)
+    clock_configure_int_divider(clk_peri,
+                    0,
+                    CLOCKS_CLK_PERI_CTRL_AUXSRC_VALUE_CLK_SYS,
+                    ROSC_HZ,
+                    OTHER_CLK_DIV);
+
+    // CLK_HSTX Transmit bit clock for the HSTX peripheral. (not used)
+    clock_configure_int_divider(clk_hstx,
+                    0,
+                    CLOCKS_CLK_HSTX_CTRL_AUXSRC_VALUE_CLK_SYS,
+                    ROSC_HZ,
+                    OTHER_CLK_DIV);
+}
+#endif
+
+
+bi_decl(bi_ptr_int32(0, 0, otp_key_page, 30));
+
+// The function lock_key() is called from decrypt() after key initialisation is complete and before decryption begins.
+// That is a suitable point to lock the OTP area where key information is stored.
+void lock_key() {
+    otp_hw->sw_lock[otp_key_page] = 0xf;
+}
+
+
+int main() {
+    // This works around E21 by locking down the page with a key,
```

Comment:
not sure what the point of this is - if using a well known key. Does this really help anyway? E21 or not, there are other errata that make it possible to run unsigned code, which would let them legitimately read OTP (unless it is page locked like this, but then where do we keep the page key?)

Reply:
This was originally just added to the encrypted bootloader example as a quick demo, and I thought I'd keep it in as it worked around E21 in a simple way. I can remove it if you'd prefer.

Reply:
Have removed it, and removed the corresponding OTP writing in the examples